scipy/spatial/transform/_rotation.pyx

```diff
@@ -223,6 +230,131 @@ cdef double[:, :] _compute_euler_from_matrix(
 
     return angles
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef double[:, :] _compute_euler_from_quat(
+    np.ndarray[double, ndim=2] quat, const uchar[:] seq, bint extrinsic=False
+):
+    # The algorithm assumes intrinsic frame transformations. The algorithm
+    # in the paper is formulated for rotation matrices which are transposition
+    # rotation matrices used within Rotation.
+    # Adapt the algorithm for our case by
+    # 1. Instead of transposing our representation, use the transpose of the
+    #    O matrix as defined in the paper, and be careful to swap indices
+    # 2. Reversing both axis sequence and angles for extrinsic rotations
+    
+    if not extrinsic:
+        seq = seq[::-1]
+
+    cdef int i = _elementary_basis_index(seq[0])
+    cdef int j = _elementary_basis_index(seq[1])
+    cdef int k = _elementary_basis_index(seq[2])
+
+    is_proper = i == k
+    if is_proper:
+        k = 3 - i - j # get third axis
+        
+    # Step 0
+    # Check if permutation is even (+1) or odd (-1)     
+    cdef int sign = int((i-j)*(j-k)*(k-i)/2)
+
+    cdef Py_ssize_t num_rotations = quat.shape[0]
+
+    # some forward definitions
+    cdef double[:, :] angles = _empty2(num_rotations, 3)
+    cdef double[:] _angles # accessor for each rotation
+    cdef double a, b, c, d, n2
+    cdef double half_sum, half_diff
+    cdef double eps = 1e-7
+    cdef bint safe1, safe2, safe, adjust
+
+    for ind in range(num_rotations):
+        _angles = angles[ind, :]
+
+        # Step 1
+        # Permutate quaternion elements            
+        if is_proper:
+            a = quat[ind, 3]
+            b = quat[ind, i]
+            c = quat[ind, j]
+            d = quat[ind, k] * sign
+        else:
+            a = quat[ind, 3] - quat[ind, j]
+            b = quat[ind, i] + quat[ind, k] * sign
+            c = quat[ind, j] + quat[ind, 3]
+            d = quat[ind, k] * sign - quat[ind, i]
+    
+        # Step 2
+        # Compute second angle...
+        n2 = a**2 + b**2 + c**2 + d**2
+        #_angles[1] = 2*acos(sqrt((a**2 + b**2) / n2))
+        _angles[1] = np.arccos(2*(a**2 + b**2) / n2 - 1)
```

Comment:
As I remember bounds violation due to roundoff issues can happen (i.e. value <-1 or >-1 for critical angle values) in a similar place for the matrix algorithm.

My suggestion is to clip the argument here.

Reply:
I think it shouldn't have a bound problem:
Since I'm dividing `a**2 + b**2` by `a**2 + b**2 + c**2 + d**2, the division will always be between 0 and 1, and the whole argument will always then be between -1 and 0, giving either `acos(-1) = pi` or `acos(1) = 0`.

Reply:
Let's see how tests are passing.

I don't know whether it can be proven that this expression (or rather the whole chain of computation considering all possible use cases) in floating point numbers never can "overflow". We have a similar situation in the matrix code, where marginal (like 1e-16 or something) overflow happened for critical angles. 

It might be different here, might be note. I think clipping is a safe and cheap option.

Reply:
I see! Interesting and yes, definitely better be safe here.